### PR TITLE
[feature] Port Export to PostgreSQL algorithm to C++

### PR DESCRIPTION
**Pull request by root676**
## Description

This PR replaces the python version of the _Export to PostgreSQL_ algorithm with a C++ port. The algorithm is now a little faster than before. :elephant: 

I am looking forward to reviews and comments!

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  mEncoding = parameterAsString( parameters, QStringLiteral( "ENCODING" ), context );
+  mOverwrite = parameterAsBoolean( parameters, QStringLiteral( "OVERWRITE" ), context );
+
+  mTable = parameterAsDatabaseTableName( parameters, QStringLiteral( "TABLENAME" ), context );
+  if ( !mTable.isEmpty() )
+  {
+    mTable = mTable.trimmed();
+  }
+  else
+  {
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:110]:
```suggestion
  mTable = parameterAsDatabaseTableName( parameters, QStringLiteral( "TABLENAME" ), context ).trimmed();
  if ( mTable.isEmpty() )
  {
```

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:110]: (The Python version is buggy here!)

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    mTable = mSource->sourceName();
+    mTable = mTable.replace( QStringLiteral( "." ), QStringLiteral( "_" ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:112]:
```suggestion
    mTable = mTable.replace( '.', '_' );
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  }
+  mTable = mTable.replace( QStringLiteral( " " ), QString() ).right( 63 );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:114]:
```suggestion
  mTable = mTable.replace( ' ', QString() ).right( 63 );
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+  mGeomColumn = parameterAsString( parameters, QStringLiteral( "GEOMETRY_COLUMN" ), context );
+  if ( mTable.isEmpty() )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:117]:
```suggestion
  if ( mGeomColumn.isEmpty() )
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    mGeomColumn = QStringLiteral( "geom" );
+  if ( mSource->wkbType() == QgsWkbTypes::NoGeometry )
+    mGeomColumn = QString(); //port note: equivalent to initializing with python None
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:120]:
```suggestion
    mGeomColumn.clear();
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+  mCreateIndex = parameterAsBoolean( parameters, QStringLiteral( "CREATEINDEX" ), context );
+
+  if ( mOverwrite )
+    mOptions[QStringLiteral( "overwrite" )] = true;
+  if ( parameterAsBoolean( parameters, QStringLiteral( "LOWERCASE_NAMES" ), context ) )
+    mOptions[QStringLiteral( "lowercaseFieldNames" )] = true;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:127]:
```suggestion
  if ( parameterAsBoolean( parameters, QStringLiteral( "LOWERCASE_NAMES" ), context ) )
  {
    mOptions[QStringLiteral( "lowercaseFieldNames" )] = true;
    mGeomColumn = mGeomColumn.toLower();  
  }
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  Q_UNUSED( parameters );
+  Q_UNUSED( context );
+
+  QgsDataSourceUri uri = QgsDataSourceUri( mConn->uri() );
+  uri.setSchema( mSchema );
+  uri.setTable( mTable );
+  uri.setKeyColumn( mPrimaryKeyField );
+  uri.setGeometryColumn( mGeomColumn );
+
+  mExporter.reset( new QgsVectorLayerExporter( uri.uri(), mProviderName, mSource->fields(), mSource->wkbType(), mSource->sourceCrs(), mOverwrite, mOptions ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:149]: This doesn't need to be a member variable
```suggestion
  std::unique_ptr< QgsVectorLayerExporter > exporter = std::make_unique< QgsVectorLayerExporter >( uri.uri(), mProviderName, mSource->fields(), mSource->wkbType(), mSource->sourceCrs(), mOverwrite, mOptions );
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+      feedback->reportError( mExporter->errorMessage() );
+
+    feedback->setProgress( i * total );
+    i++;
+  }
+  mExporter->flushBuffer();
+
+  if ( mExporter->errorCode() != Qgis::VectorExportResult::Success )
+    throw QgsProcessingException( QObject::tr( "Error importing into PostGIS\n%1" ).arg( mExporter->errorMessage() ) );
+
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:175]:
```suggestion
 exporter.reset();
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  QgsDataSourceUri uri = QgsDataSourceUri( mConn->uri() );
+  uri.setSchema( mSchema );
+  uri.setTable( mTable );
+  uri.setKeyColumn( mPrimaryKeyField );
+  uri.setGeometryColumn( mGeomColumn );
+
+  mExporter.reset( new QgsVectorLayerExporter( uri.uri(), mProviderName, mSource->fields(), mSource->wkbType(), mSource->sourceCrs(), mOverwrite, mOptions ) );
+
+  if ( mExporter->errorCode() != Qgis::VectorExportResult::Success )
+    throw QgsProcessingException( QObject::tr( "Error importing into PostGIS\n%1" ).arg( mExporter->errorMessage() ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:152]:
```suggestion
    throw QgsProcessingException( QObject::tr( "Error exporting to PostGIS\n%1" ).arg( mExporter->errorMessage() ) );
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+  QgsFeatureIterator featureIterator = mSource->getFeatures();
+
+  double total = ( mSource->featureCount() ) ? 100.0 / mSource->featureCount() : 0.0;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:156]:
```suggestion
  const double progressStep = ( mSource->featureCount() ) ? 100.0 / mSource->featureCount() : 0.0;
```

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    if ( !mExporter->addFeature( f, QgsFeatureSink::FastInsert ) )
+      feedback->reportError( mExporter->errorMessage() );
+
+    feedback->setProgress( i * total );
+    i++;
+  }
+  mExporter->flushBuffer();
+
+  if ( mExporter->errorCode() != Qgis::VectorExportResult::Success )
+    throw QgsProcessingException( QObject::tr( "Error importing into PostGIS\n%1" ).arg( mExporter->errorMessage() ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:174]:
```suggestion
    throw QgsProcessingException( QObject::tr( "Error exporting to PostGIS\n%1" ).arg( mExporter->errorMessage() ) );
```

**Comment by nyalldawson** [comment]
Nice work @root676 ! Just some minor changes required here...

**Comment by root676** [comment]
@nyalldawson thank you and thanks for the super fast review! :rocket:  I think everything has been addressed now.

File: src/analysis/processing/qgsalgorithmexporttopostgresql.cpp [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  if ( mSource->wkbType() == QgsWkbTypes::NoGeometry )
+    mGeomColumn.clear();
+
+  mCreateIndex = parameterAsBoolean( parameters, QStringLiteral( "CREATEINDEX" ), context );
+
+  if ( mOverwrite )
+    mOptions[QStringLiteral( "overwrite" )] = true;
+  if ( parameterAsBoolean( parameters, QStringLiteral( "LOWERCASE_NAMES" ), context ) )
+    mOptions[QStringLiteral( "lowercaseFieldNames" )] = true;
+  mGeomColumn = mGeomColumn.toLower();
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmexporttopostgresql.cpp:129]:
```suggestion
  if ( parameterAsBoolean( parameters, QStringLiteral( "LOWERCASE_NAMES" ), context ) )
  {
    mOptions[QStringLiteral( "lowercaseFieldNames" )] = true;
    mGeomColumn = mGeomColumn.toLower();
  }
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
@root676 there's a related test failure here -- can you investigate?

**Pull request merged**